src/wp-includes/user.php

```diff
@@ -3012,11 +3012,56 @@ function retrieve_password( $user_login = null ) {
 	 */
 	$message = apply_filters( 'retrieve_password_message', $message, $key, $user_login, $user_data );
 
+	// Wrap the single notification email arguments in an array to pass them to the retrieve_password_notification_email filter.
+	$defaults = array(
+		'to'      => $user_email,
+		'subject' => $title,
+		'message' => $message,
+		'headers' => '',
+	);
+
+	$data = compact( 'key', 'user_login', 'user_data' );
+
+	/**
+	 * Filter the contents of the reset password notification email sent to the user. Return false to prevent sending the email.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @param array   $defaults {
+	 *     The default notification email arguments. Used to build wp_mail().
+	 *
+	 *     @type string $to      The intended recipient - user email address.
+	 *     @type string $subject The subject of the email.
+	 *     @type string $message The body of the email.
+	 *     @type string $headers The headers of the email.
+	 * }
+	 * @param array $data {
+	 *     Additional information for extenders.
+	 *
+	 *     @type string  $key        The activation key.
+	 *     @type string  $user_login The username for the user.
+	 *     @type WP_User $user_data  WP_User object.
+	 * }
+	 */
+	$notification_email = apply_filters( 'retrieve_password_notification_email', $defaults, $data );
+
 	if ( $switched_locale ) {
 		restore_previous_locale();
 	}
 
-	if ( $message && ! wp_mail( $user_email, wp_specialchars_decode( $title ), $message ) ) {
+	// Prevent sending the reset password notification email if the filtered value is false.
+	if ( is_array( $notification_email ) ) {
+		// Merge defaults with the filtered array as fallback in case any required value is missing in the filtered array.
```

Comment:
```suggestion
		// Force key order and merge defaults in case any value is missing in the filtered array.
```
Added "Force key order" as another important benefit of using `array_merge()` here and made the comment more concise.